src/EditorFeatures/Core/Tagging/CompilationAvailableTaggerEventSource.cs

```diff
@@ -1,89 +0,0 @@
-using Microsoft.CodeAnalysis.Host;
-using Microsoft.CodeAnalysis.Remote;
-using Microsoft.CodeAnalysis.Shared.TestHooks;
-using Microsoft.CodeAnalysis.Shared.Utilities;
-using Microsoft.CodeAnalysis.Tagging;
-using Microsoft.CodeAnalysis.Text;
-using Microsoft.VisualStudio.Text;
-using Microsoft.VisualStudio.Threading;
-using Roslyn.Utilities;
-
-namespace Microsoft.CodeAnalysis.Editor.Tagging;
-
-/// <summary>
-/// Tagger event that fires once the compilation is available in the remote OOP process for a particular project.
-/// Used to trigger things such as:
-/// <list type="bullet">
-/// <item>reclassification pass as classification may show either cached classifications (from a  previous session),
-/// or incomplete classifications due to frozen-partial compilations being used.</item>
-/// <item>recomputation of navigation bar items due to frozen-partial compilations being used.</item>
-/// <item>recomputation of inheritance margin items due to frozen-partial compilations being used.</item>
-/// </list>
-/// </summary>
-internal sealed class CompilationAvailableTaggerEventSource : ITaggerEventSource
-{
-    private readonly ITextBuffer _subjectBuffer;
-
-    /// <summary>
-    /// Other event sources we're composing over.  If they fire, we should reclassify.  However, after they fire, we
-    /// should also refire an event once we get the next full compilation ready.
-    /// </summary>
-    private readonly ITaggerEventSource _underlyingSource;
-
-    private readonly CompilationAvailableEventSource _eventSource;
-
-    private readonly Action _onCompilationAvailable;
-
-    public CompilationAvailableTaggerEventSource(
-        ITextBuffer subjectBuffer,
-        IAsynchronousOperationListener asyncListener,
-        params ITaggerEventSource[] eventSources)
-    {
-        _subjectBuffer = subjectBuffer;
-        _eventSource = new CompilationAvailableEventSource(asyncListener);
-        _underlyingSource = TaggerEventSources.Compose(eventSources);
-        _onCompilationAvailable = () => this.Changed?.Invoke(this, TaggerEventArgs.Empty);
-    }
-
-    public event EventHandler<TaggerEventArgs>? Changed;
-
-    public void Connect()
-    {
-        // When we are connected to, connect to all our underlying sources and have them notify us when they've changed.
-        _underlyingSource.Connect();
-        _underlyingSource.Changed += OnUnderlyingSourceChanged;
-    }
-
-    public void Disconnect()
-    {
-        _underlyingSource.Changed -= OnUnderlyingSourceChanged;
-        _underlyingSource.Disconnect();
-        _eventSource.Dispose();
-    }
-
-    public void Pause()
-        => _underlyingSource.Pause();
-
-    public void Resume()
-        => _underlyingSource.Resume();
-
-    private void OnUnderlyingSourceChanged(object? sender, TaggerEventArgs args)
-    {
-        // First, notify anyone listening to us that something definitely changed.
-        this.Changed?.Invoke(this, args);
```

Comment:
>this.Changed?.Invoke(this, args); [](http://example.com/codeflow?start=8&length=33)

I think I'm clear on not notifiying listeners on frozen partial compilations, but would it make sense to still notify them when the compilation has actually reached a final state?

Reply:
so the main problem is simply that we'd need those features to have a new concept (Which doesn't exist at all).  Which is:

1. use the compilation if available, but otherwise, use teh frozen partial one.  

Note that the above is highly problematic as well as the features will often be working with a slightly changed doc, and thus are calling "get frozen partial snapshot with changes".  In that world, we def do not want anything related to the real compilation as that involves computing generators.  

This basically also involves the workspace having some sort of internal eventing model itself for these sorts of concepts.  I'm *very* wary of that, esp. as it encourages more of a 'push' approach for these features.  

My preference for now is simply saying: if you use frozen partial, you are opting into staleness against the snapshot you're currently operating on.